File: bin/lib/lakera.js
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Lakera Guard integration — screens inference requests/responses for prompt
// injection, jailbreaks, PII leakage, and content violations.
//
// Supports two modes:
//   1. SaaS API  — calls api.lakera.ai (requires LAKERA_GUARD_API_KEY)
//   2. Sidecar   — local container on localhost:8932 (self-hosted)
//
// Env:
//   LAKERA_GUARD_API_KEY   API key for Lakera Guard SaaS
//   LAKERA_GUARD_URL       Override guard endpoint (default: https://api.lakera.ai)
//   LAKERA_GUARD_MODE      "saas" | "sidecar" | "disabled" (default: "saas" if key present)

const { runCapture, run } = require("./runner");

const LAKERA_SIDECAR_IMAGE = "lakera/lakera-guard:latest";
const LAKERA_SIDECAR_PORT = 8932;

/**
 * Determine the current guard mode based on environment variables.
 */
function resolveMode() {
  const explicit = process.env.LAKERA_GUARD_MODE;
  if (explicit) return explicit;
  if (process.env.LAKERA_GUARD_API_KEY) return "saas";
  return "disabled";
}

/**
 * Return the base URL for the Lakera Guard API.
 */
function guardBaseUrl() {
  if (process.env.LAKERA_GUARD_URL) {
    return process.env.LAKERA_GUARD_URL.replace(/\/$/, "");
  }
  const mode = resolveMode();
  if (mode === "sidecar") return `http://127.0.0.1:${LAKERA_SIDECAR_PORT}`;
  return "https://api.lakera.ai";
}

/**
 * Start the Lakera Guard sidecar container.
 * Returns { ok, port } on success.
 */
function startSidecar() {
  const name = "lakera-guard";

  // Stop any existing container
  run(`docker rm -f ${name} 2>/dev/null || true`, { ignoreError: true });

  const apiKey = process.env.LAKERA_GUARD_API_KEY || "";
  const { execFileSync } = require("child_process");
  const dockerArgs = [
    "run", "-d", "--name", name, "--restart", "unless-stopped",
    "-p", `${LAKERA_SIDECAR_PORT}:8000`,
  ];
  if (apiKey) {
    dockerArgs.push("-e", `LAKERA_GUARD_API_KEY=${apiKey}`);
  }
  dockerArgs.push(LAKERA_SIDECAR_IMAGE);
  execFileSync("docker", dockerArgs, { stdio: "pipe" });

  // Wait for health
  for (let i = 0; i < 15; i++) {
    const health = runCapture(
      `curl -sf http://127.0.0.1:${LAKERA_SIDECAR_PORT}/health 2>/dev/null`,
      { ignoreError: true }
    );
    if (health && health.includes("ok")) {
      return { ok: true, port: LAKERA_SIDECAR_PORT };
    }
    require("child_process").spawnSync("sleep", ["2"]);
  }

  return { ok: false, port: LAKERA_SIDECAR_PORT };
}

/**
 * Stop and remove the sidecar container.
 */
function stopSidecar() {
  run("docker rm -f lakera-guard 2>/dev/null || true", { ignoreError: true });
}

/**
 * Check if the sidecar container is running.
 */
function sidecarStatus() {
  const out = runCapture(
    'docker inspect --format="{{.State.Running}}" lakera-guard 2>/dev/null',
    { ignoreError: true }
  );
  return { running: out && out.trim() === "true" };
}

/**
 * Screen a prompt (pre-inference) via Lakera Guard API.
 *
 * @param {Array<{role: string, content: string}>} messages - Chat messages
 * @returns {{ flagged: boolean, categories: string[], raw: object }}
 */
function screenRequest(messages) {
  const mode = resolveMode();
  if (mode === "disabled") return { flagged: false, categories: [], raw: {} };

  const baseUrl = guardBaseUrl();
  const apiKey = process.env.LAKERA_GUARD_API_KEY || "";
  const authHeader = apiKey ? `-H "Authorization: Bearer ${apiKey}"` : "";

  const payload = JSON.stringify({ messages });
  // Use a temp file to avoid shell escaping issues with message content
  const fs = require("fs");
  const os = require("os");
  const path = require("path");
  const tmpFile = path.join(os.tmpdir(), `lakera-req-${Date.now()}.json`);
  fs.writeFileSync(tmpFile, payload, "utf8");

  try {
    const result = runCapture(
      `curl -sf -X POST "${baseUrl}/v2/guard" ` +
      `${authHeader} ` +
      `-H "Content-Type: application/json" ` +
      `-d @"${tmpFile}" 2>/dev/null`,
      { ignoreError: true }
    );

    if (!result) return { flagged: false, categories: [], raw: {} };

    const parsed = JSON.parse(result);
    const categories = (parsed.flagged_categories || []).map((c) => c.category || c);
    return {
      flagged: categories.length > 0,
      categories,
      raw: parsed,
    };
  } catch {
    // Guard unavailable — fail open (log but don't block)
    return { flagged: false, categories: [], raw: { error: "guard_unavailable" } };
  } finally {
    try { require("fs").unlinkSync(tmpFile); } catch { /* ignore */ }
  }
}

/**
 * Screen an LLM response (post-inference) via Lakera Guard API.
 *
 * @param {string} responseText - The LLM's response text
 * @param {Array<{role: string, content: string}>} context - Original messages for context
 * @returns {{ flagged: boolean, categories: string[], raw: object }}
 */
function screenResponse(responseText, context) {
  const fullMessages = [
    ...(context || []),
    { role: "assistant", content: responseText },
  ];
  return screenRequest(fullMessages);
}

module.exports = {
  resolveMode,
  guardBaseUrl,
  startSidecar,
  stopSidecar,
  sidecarStatus,
  screenRequest,
  screenResponse,
  LAKERA_SIDECAR_PORT,
};

File: nemoclaw-blueprint/orchestrator/lakera.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""
Lakera Guard integration for the NemoClaw blueprint runner.

Screens inference requests and responses for prompt injection, jailbreaks,
PII leakage, and content violations via the Lakera Guard API.

Modes:
  saas     — calls api.lakera.ai (requires LAKERA_GUARD_API_KEY env)
  sidecar  — local container on localhost:8932
  disabled — no screening (default when no key is set)

Env:
  LAKERA_GUARD_API_KEY   API key for Lakera Guard SaaS
  LAKERA_GUARD_URL       Override guard endpoint
  LAKERA_GUARD_MODE      "saas" | "sidecar" | "disabled"
"""

from __future__ import annotations

import json
import logging
import os
import subprocess
import tempfile
import time
from dataclasses import dataclass, field
from typing import Any

logger = logging.getLogger(__name__)

LAKERA_SIDECAR_IMAGE = "lakera/lakera-guard:latest"
LAKERA_SIDECAR_PORT = 8932


@dataclass
class ScreenResult:
    """Result from a Lakera Guard screening call."""

    flagged: bool = False
    categories: list[str] = field(default_factory=list)
    raw: dict[str, Any] = field(default_factory=dict)


def resolve_mode() -> str:
    """Determine the guard mode from environment."""
    explicit = os.environ.get("LAKERA_GUARD_MODE", "")
    if explicit:
        return explicit
    if os.environ.get("LAKERA_GUARD_API_KEY"):
        return "saas"
    return "disabled"


def guard_base_url() -> str:
    """Return the Lakera Guard API base URL."""
    override = os.environ.get("LAKERA_GUARD_URL", "")
    if override:
        return override.rstrip("/")
    mode = resolve_mode()
    if mode == "sidecar":
        return f"http://127.0.0.1:{LAKERA_SIDECAR_PORT}"
    return "https://api.lakera.ai"


def start_sidecar() -> dict[str, Any]:
    """Start the Lakera Guard sidecar Docker container."""
    name = "lakera-guard"
    subprocess.run(
        ["docker", "rm", "-f", name],
        capture_output=True,
        check=False,
    )

    api_key = os.environ.get("LAKERA_GUARD_API_KEY", "")
    cmd = [
        "docker", "run", "-d",
        "--name", name,
        "--restart", "unless-stopped",
        "-p", f"{LAKERA_SIDECAR_PORT}:8000",
    ]
    if api_key:
        cmd.extend(["-e", f"LAKERA_GUARD_API_KEY={api_key}"])
    cmd.append(LAKERA_SIDECAR_IMAGE)

    subprocess.run(cmd, check=True, capture_output=True, text=True)

    # Wait for health
    for _ in range(15):
        try:
            result = subprocess.run(
                ["curl", "-sf", f"http://127.0.0.1:{LAKERA_SIDECAR_PORT}/health"],
                capture_output=True,
                text=True,
                check=False,
            )
            if result.returncode == 0 and "ok" in result.stdout.lower():
                return {"ok": True, "port": LAKERA_SIDECAR_PORT}
        except Exception:
            pass
        time.sleep(2)

    return {"ok": False, "port": LAKERA_SIDECAR_PORT}


def stop_sidecar() -> None:
    """Stop and remove the sidecar container."""
    subprocess.run(
        ["docker", "rm", "-f", "lakera-guard"],
        capture_output=True,
        check=False,
    )


def screen_messages(messages: list[dict[str, str]]) -> ScreenResult:
    """
    Screen a list of chat messages via Lakera Guard.

    Args:
        messages: List of {role, content} dicts (OpenAI chat format).

    Returns:
        ScreenResult with flagged status, categories, and raw API response.
    """
    mode = resolve_mode()
    if mode == "disabled":
        return ScreenResult()

    base_url = guard_base_url()
    api_key = os.environ.get("LAKERA_GUARD_API_KEY", "")

    payload = json.dumps({"messages": messages})

    with tempfile.NamedTemporaryFile(
        mode="w", suffix=".json", delete=False
    ) as tmp:
        tmp.write(payload)
        tmp_path = tmp.name

    try:
        curl_cmd = [
            "curl", "-sf", "-X", "POST",
            f"{base_url}/v2/guard",
            "-H", "Content-Type: application/json",
        ]
        if api_key:
            curl_cmd.extend(["-H", f"Authorization: Bearer {api_key}"])
        curl_cmd.extend(["-d", f"@{tmp_path}"])

        result = subprocess.run(
            curl_cmd,
            capture_output=True,
            text=True,
            check=False,
            timeout=10,
        )

        if result.returncode != 0 or not result.stdout.strip():
            logger.warning("Lakera Guard unavailable — failing open")
            return ScreenResult(raw={"error": "guard_unavailable"})

        parsed = json.loads(result.stdout)
        categories = [
            c.get("category", c) if isinstance(c, dict) else c
            for c in parsed.get("flagged_categories", [])
        ]
        return ScreenResult(
            flagged=len(categories) > 0,
            categories=categories,
            raw=parsed,
        )
    except (json.JSONDecodeError, subprocess.TimeoutExpired) as exc:
        logger.warning("Lakera Guard error — failing open: %s", exc)
        return ScreenResult(raw={"error": str(exc)})
    finally:
        try:
            os.unlink(tmp_path)
        except OSError:
            pass


def screen_request(
    system_prompt: str,
    user_message: str,
) -> ScreenResult:
    """Screen a pre-inference request (system prompt + user message)."""
    messages = []
    if system_prompt:
        messages.append({"role": "system", "content": system_prompt})
    messages.append({"role": "user", "content": user_message})
    return screen_messages(messages)


def screen_response(
    response_text: str,
    context: list[dict[str, str]] | None = None,
) -> ScreenResult:
    """Screen a post-inference response."""
    messages = list(context or [])
    messages.append({"role": "assistant", "content": response_text})
    return screen_messages(messages)


def guard_config_for_plan(blueprint: dict[str, Any]) -> dict[str, Any]:
    """
    Extract Lakera Guard configuration from the blueprint for inclusion
    in a deployment plan.
    """
    guard_cfg = (
        blueprint.get("components", {}).get("guard", {})
    )
    mode = resolve_mode()
    return {
        "enabled": mode != "disabled",
        "mode": mode,
        "endpoint": guard_base_url() if mode != "disabled" else None,
        "fail_policy": guard_cfg.get("fail_policy", "open"),
        "screen_requests": guard_cfg.get("screen_requests", True),
        "screen_responses": guard_cfg.get("screen_responses", True),
    }
